src/s3ql/backends/gdrive.py

```diff
@@ -0,0 +1,599 @@
+        if len(client_secret_and_refresh_token) != 2:
+            raise AuthenticationError("Invalid Password format, must be  secretApplication:refreshToken")
+        self.client_secret = client_secret_and_refresh_token[0]
+        self.refresh_token = client_secret_and_refresh_token[1]
+        self.options = options
+        self.delete_map_state = dict()
+
+        # get google drive service
+        global credentials        
+        global mainFolder
+        if credentials == None:
+            credentials = self._get_credentials()
+        http = credentials.authorize(httplib2.Http())        
+
+        self.service = apiclient.discovery.build('drive', 'v3', http=http,cache_discovery=False)
+
+        # get gdrive folder
+        if mainFolder==None:
+            folderPath = storage_url[len('gdrive://'):].rstrip('/')
+            self.folder = self._lookup_file(folderPath)
+            if self.folder != None:
+                if self.folder['mimeType'] != Backend.MIME_TYPE_FOLDER:
+                    raise DanglingStorageURLError(folderPath)
+            else:
+                self.folder = self.service.files().create(body={'name' : folderPath,'mimeType' : Backend.MIME_TYPE_FOLDER},fields='id').execute()
+            mainFolder=self.folder
+        else:
+            self.folder = mainFolder
+
+    def _get_credentials(self):
+        return oauth2client.client.OAuth2Credentials(access_token="1234",client_id=self.login,client_secret=self.client_secret,
+        refresh_token=self.refresh_token,token_expiry="0000-00-00T00:00:00Z",token_uri=Backend.TOKEN_URI,user_agent=Backend.USER_AGENT)
+
+    @copy_ancestor_docstring
+    def is_temp_failure(self, exc): #IGNORE:W0613
+        if isinstance(exc, apiclient.errors.HttpError):
+            #handle only in case response are:
+            if exc.resp.status == 403 and exc.resp.reason in ["Forbidden","userRateLimitExceeded","rateLimitExceeded","backendError"]:                
+                return True
+            #Google best practice error says error 500 could be temporal, also We get some resumable errors with 400 Bad Request (Fully Random)
+            elif exc.resp.status >= 500 or exc.resp.status == 400:
+                return True
+        elif isinstance(exc,(BrokenPipeError,ConnectionResetError,socket.gaierror,socket.herror,socket.timeout,http.client.IncompleteRead,
+        httplib2.ServerNotFoundError,ssl.SSLZeroReturnError, ssl.SSLEOFError,ssl.SSLSyscallError,ssl.SSLError)):
+            return True
+        return False
+
+    @staticmethod
+    def _escape_string(name):
+        '''Escape string'''
+        return name.replace("\\", "\\\\").replace("'", "\\'")
+
+    def _list_files(self, folder, fields, query=None):
+        '''List folder contents
+        Arguments:
+        folder -- Google Drive folder to list files in
+        fields -- file properties to return
+        query -- files.list query (optional)
```

Comment:
I do not understand what *query* is for

Reply:
google drive have their own "SQL" to find files so this var allow you to use  gdrive query language  to find files.

Reply:
Could you please put something along these lines in the docstring? A link to the webpage that describes the query language would be nice too.